bot/exts/info/patreon.py

```diff
@@ -0,0 +1,128 @@
+
+from bot import constants, utils
+from bot.bot import Bot
+from bot.log import get_logger
+from bot.utils.channel import get_or_fetch_channel
+from bot.utils.members import has_role_id
+
+log = get_logger(__name__)
+
+PATREON_INFORMATION = (
+    "Python Discord is not-for-profit and volunteer run, so we rely on Patreon donations to do what we do. "
+    "We use the money we get to offer excellent prizes for all of our events. Prizes like t-shirts, "
+    "stickers, microcontrollers that support CircuitPython, or maybe even a mechanical keyboard.\n\n"
+    "You can read more about how Patreon donations help us, and consider donating yourself, on our patreon page "
+    "[here](https://pydis.com/patreon)!"
+)
+
+# List of tuples containing tier number and Discord role ID.
+# Ordered from highest tier to lowest.
+PATREON_TIERS: list[tuple[int, int]] = [
+    (3, constants.Roles.patreon_tier_3),
+    (2, constants.Roles.patreon_tier_2),
+    (1, constants.Roles.patreon_tier_1),
+]
+
+
+def get_patreon_tier(member: discord.Member) -> int:
+    """
+    Get the patreon tier of `member`.
+
+    A patreon tier of 0 indicates the user is not a patreon.
+    """
+    for tier, role_id in PATREON_TIERS:
+        if has_role_id(member, role_id):
+            return tier
+    return 0
+
+
+class Patreon(commands.Cog):
+    """Cog that shows patreon supporters."""
+
+    def __init__(self, bot: Bot) -> None:
+        self.bot = bot
+
+        self.current_monthly_supporters.start()
+
+    @commands.Cog.listener()
+    async def on_member_update(self, before: discord.Member, after: discord.Member) -> None:
+        """Send a message when someone receives a patreon role."""
+        old_patreon_tier = get_patreon_tier(before)
+        new_patreon_tier = get_patreon_tier(after)
+
+        if not new_patreon_tier > old_patreon_tier:
+            return
+
+        message = (
+            f":tada: {after.mention} just became a **tier {new_patreon_tier}** patron!\n"
+            "[Support us on Patreon](https://pydis.com/patreon)"
+        )
+        channel = await utils.channel.get_or_fetch_channel(constants.Channels.meta)
+        role = after.guild.get_role(new_patreon_tier)
+
+        await channel.send(
+            embed=discord.Embed(
+                description=message,
+                colour=role.colour,
+            )
+        )
+
+    async def send_current_supporters(self, channel: discord.abc.Messageable) -> None:
+        """Send the current list of patreon supporters, sorted by tier level."""
+        guild = self.bot.get_guild(constants.Guild.id)
+
+        embed_list = []
+        for tier, role_id in PATREON_TIERS:
+            role = guild.get_role(role_id)
+
+            # Filter out any members where this is not their highest tier.
+            patrons = [member for member in role.members if get_patreon_tier(member) == tier]
+            patron_names = [f"{patron.mention} ({patron.name}#{patron.discriminator})" for patron in patrons]
```

Comment:
I'm not sure how well having mentions in embeds will work, given they often don't resolve. Would it be better to make this a normal text message with allowed_mentions to prevent mentioning users. The character limit is much more limiting unless we use 3 separate messages though.

I think I should also add some sort of overflow handling if there are too many users. I'm thinking I might shuffle each tier and pick the first 10 or something.

Reply:
This is how it currently looks on server for me (ignore first embed):

![image](https://user-images.githubusercontent.com/22353562/185756136-b8077ea3-5254-49ce-baf7-6bb68c034438.png)

Not great. I'll try with a normal text message

Reply:
Since you're likely going to remove the mentions, why not list all the users in paragraph form? Then we shouldn't need to pick only a random 10.

Reply:
Should we consider maybe multi-column tables or embeds? Should at least be able to fit in all the patron supporters that way.

![img1](https://i.imgur.com/DUPQ6HX.png)

![img2](https://i.imgur.com/SaWBMLk.png)

Reply:
I've gone with plain text names listed in the embeds, I think that's neatest.